cpu.cpp

```diff
@@ -560,6 +697,1224 @@ void Cpu::sub()
+    switch(cgb->getAndIncrementImpulse()) {
+    case 1:
+        decideNextPhase();
+        qDebug() << "EX NOP I1";
+
+        break;
+    default:
+        qDebug() << "ERROR EX NOP";
+        break;
+    }
+}
+
+void Cpu::ret() {
+    switch(cgb->getAndIncrementImpulse()) {
+    case 1:
+        SBUS = SP;
+        emit PdSPS(true);
+
+        RBUS = SBUS;
+        emit ALU(true, true, false, "SBUS");
+        emit PdALU(true);
+
+        ADR = RBUS;
+        emit PmADR(true, ADR);
+
+        qDebug() << "EX RET I1";
+
+        break;
+    case 2:
+        MDR = (memory[ADR + 1] << 8) | memory[ADR];
+        emit RD(true, "READ");
+        emit PmMDR(true, MDR);
+
+        qDebug() << "EX RET I2";
+
+        break;
+    case 3:
+        DBUS = MDR;
+        emit PdMDRD(true);
+
+        RBUS = DBUS;
+        emit ALU(true, false, true, "DBUS");
+        emit PdALU(true);
+
+        PC = RBUS;
+        emit PmPC(true, PC);
+
+        SP += 2;
+        emit SPchanged(true, SP);
+
+        decideNextPhase();
+
+        qDebug() << "EX RET I3";
+
+        break;
+    default:
+        qDebug() << "ERROR EX RET";
+        break;
+    }
+}
+
+void Cpu::reti() {
+    switch(cgb->getAndIncrementImpulse()) {
+    case 1:
+        intr = false;
+
+        SBUS = SP;
+        emit PdSPS(true);
+
+        RBUS = SBUS;
+        emit ALU(true, true, false, "SBUS");
+        emit PdALU(true);
+
+        ADR = RBUS;
+        emit PmADR(true, ADR);
+
+        qDebug() << "EX RETI I1";
+
+        break;
+    case 2:
+        MDR = (memory[ADR + 1] << 8) | memory[ADR];
+        emit RD(true, "READ");
+        emit PmMDR(true, MDR);
+
+        qDebug() << "EX RETI I2";
+
+        break;
+    case 3:
+        DBUS = MDR;
+        emit PdMDRD(true);
+
+        RBUS = DBUS;
+        emit ALU(true, false, true, "DBUS");
+        emit PdALU(true);
+
+        PC = RBUS;
+        emit PmPC(true, PC);
+
+        SP += 2;
+        emit SPchanged(true, SP);
+
+        qDebug() << "EX RETI I3";
+
+        break;
+    case 4:
+        SBUS = SP;
+        emit PdSPS(true);
+
+        RBUS = SBUS;
+        emit ALU(true, true, false, "SBUS");
+        emit PdALU(true);
+
+        ADR = RBUS;
+        emit PmADR(true, ADR);
+
+        qDebug() << "EX RETI I4";
+
+        break;
+    case 5:
+        MDR = (memory[ADR + 1] << 8) | memory[ADR];
+        emit RD(true, "READ");
+        emit PmMDR(true, MDR);
+
+        qDebug() << "EX RETI I5";
+
+        break;
+    case 6:
+        DBUS = MDR;
+        emit PdMDRD(true);
+
+        RBUS = DBUS;
+        emit ALU(true, false, true, "DBUS");
+        emit PdALU(true);
+
+        FLAG = RBUS;
+        emit PmFLAG(true, FLAG, true);
+
+        SP += 2;
+        emit SPchanged(true, SP);
+
+        decideNextPhase();
+
+        qDebug() << "EX RETI I6";
+
+        break;
+    default:
+        qDebug() << "ERROR EX RETI";
+    }
+}
+
+void Cpu::pushpc() {
+    switch(cgb->getAndIncrementImpulse()) {
+    case 1:
+        SP -= 2;
```

Comment:
What happens if this is the first instruction in the program? Or when sp goes negative?